File: custom_kernel/src/drivers/rtc.rs
use core::arch::asm;

const CMOS_ADDR: u16 = 0x70;
const CMOS_DATA: u16 = 0x71;

pub struct RtcTime {
    pub seconds: u8,
    pub minutes: u8,
    pub hours: u8,
    pub day: u8,
    pub month: u8,
    pub year: usize,
}

use core::sync::atomic::{AtomicI32, Ordering};
pub static TIMEZONE_OFFSET_HOURS: AtomicI32 = AtomicI32::new(0);

pub fn init() {
    // RTC is always present on x86, no init needed. 
    // Kept for API compatibility.
}

fn read_register(reg: u8) -> u8 {
    unsafe {
        asm!("out dx, al", in("dx") CMOS_ADDR, in("al") reg, options(nomem, nostack, preserves_flags));
        let mut ret: u8;
        asm!("in al, dx", out("al") ret, in("dx") CMOS_DATA, options(nomem, nostack, preserves_flags));
        ret
    }
}

fn bcd_to_binary(bcd: u8) -> u8 {
    (bcd & 0x0F) + ((bcd / 16) * 10)
}


fn get_update_in_progress_flag() -> bool {
    unsafe {
        asm!("out dx, al", in("dx") CMOS_ADDR, in("al") 0x0Au8, options(nomem, nostack, preserves_flags));
        let mut ret: u8;
        asm!("in al, dx", out("al") ret, in("dx") CMOS_DATA, options(nomem, nostack, preserves_flags));
        (ret & 0x80) != 0
    }
}

pub fn read_time() -> RtcTime {
    let mut t = RtcTime { seconds: 0, minutes: 0, hours: 0, day: 0, month: 0, year: 0 };
    let mut last = RtcTime { seconds: 0, minutes: 0, hours: 0, day: 0, month: 0, year: 0 };

    loop {
        // Wait until update is not in progress
        while get_update_in_progress_flag() {
             core::hint::spin_loop();
        }

        t.seconds = read_register(0x00);
        t.minutes = read_register(0x02);
        t.hours = read_register(0x04);
        t.day = read_register(0x07);
        t.month = read_register(0x08);
        t.year = read_register(0x09) as usize;

        // Verify stability (read twice)
         while get_update_in_progress_flag() {
             core::hint::spin_loop();
        }
        
        last.seconds = read_register(0x00);
        last.minutes = read_register(0x02);
        last.hours = read_register(0x04);
        last.day = read_register(0x07);
        last.month = read_register(0x08);
        last.year = read_register(0x09) as usize;

        if t.seconds == last.seconds && t.minutes == last.minutes && t.hours == last.hours &&
           t.day == last.day && t.month == last.month && t.year == last.year {
            break;
        }
    }

    let register_b = read_register(0x0B);

    // Convert BCD to binary if needed
    if (register_b & 0x04) == 0 {
        t.seconds = bcd_to_binary(t.seconds);
        t.minutes = bcd_to_binary(t.minutes);
        // Handle 12-hour format bit (0x80) if present in BCD mode before masking? 
        // Usually Hour is BCD encoded 0-23 or 1-12. 
        // 0x80 bit usually means PM in 12h mode.
        // Let's safe-guard:
        t.hours = bcd_to_binary(t.hours & 0x7F); 
        t.day = bcd_to_binary(t.day);
        t.month = bcd_to_binary(t.month);
        t.year = bcd_to_binary(t.year as u8) as usize;
    }

    // Convert 12 hour to 24 hour if necessary
    // Bit 1 of Reg B: 1 = 24h, 0 = 12h
    // In 12h mode, bit 7 of hour byte determines PM.
    if (register_b & 0x02) == 0 && (t.hours & 0x80) != 0 {
         t.hours = ((t.hours & 0x7F) + 12) % 24;
    }
    
    // Convert year to full year (assume 20xx)
    let full_year = 2000 + t.year;
    t.year = full_year;

    // Apply Timezone Offset
    let offset = TIMEZONE_OFFSET_HOURS.load(Ordering::Relaxed);
    if offset != 0 {
        let mut h = t.hours as i32 + offset;
        if h < 0 {
            h += 24;
            // Simplistic day wrap (doesn't handle month/year boundaries properly, good enough for toy OS)
            if t.day > 1 { t.day -= 1; }
        } else if h >= 24 {
            h -= 24;
            t.day += 1; // Simplistic day wrap
        }
        t.hours = h as u8;
    }
    
    t
}

File: custom_kernel/src/c/drivers/rtc.c
#include "rtc.h"
#include "kernel/io.h" /* for outb, inb */

#define CMOS_ADDRESS 0x70
#define CMOS_DATA    0x71

static int bcd2bin(int bcd) {
    return ((bcd >> 4) * 10) + (bcd & 0x0F);
}

static uint8_t read_register(uint8_t reg) {
    outb(CMOS_ADDRESS, reg);
    return inb(CMOS_DATA);
}

static int get_update_in_progress_flag() {
    outb(CMOS_ADDRESS, 0x0A);
    return (inb(CMOS_DATA) & 0x80);
}

void rtc_init(void) {
    /* No init technically required for basic reading */
}

void rtc_get_time(rtc_time_t *t) {
    uint8_t last_second;
    uint8_t last_minute;
    uint8_t last_hour;
    uint8_t last_day;
    uint8_t last_month;
    uint8_t last_year;
    uint8_t century;
    uint8_t registerB;

    /* Wait for update to complete */
    while (get_update_in_progress_flag());

    t->second = read_register(0x00);
    t->minute = read_register(0x02);
    t->hour = read_register(0x04);
    t->day = read_register(0x07);
    t->month = read_register(0x08);
    t->year = read_register(0x09);
    
    /* Optional: Centure register? usually not standard. assume 20xx */
    
    do {
        last_second = t->second;
        last_minute = t->minute;
        last_hour = t->hour;
        last_day = t->day;
        last_month = t->month;
        last_year = t->year;

        while (get_update_in_progress_flag());

        t->second = read_register(0x00);
        t->minute = read_register(0x02);
        t->hour = read_register(0x04);
        t->day = read_register(0x07);
        t->month = read_register(0x08);
        t->year = read_register(0x09);
    } while ((last_second != t->second) || (last_minute != t->minute) || (last_hour != t->hour) ||
               (last_day != t->day)     || (last_month != t->month)   || (last_year != t->year));

    registerB = read_register(0x0B);

    /* Convert BCD to binary values if necessary */
    if (!(registerB & 0x04)) {
        t->second = bcd2bin(t->second);
        t->minute = bcd2bin(t->minute);
        t->hour   = bcd2bin(t->hour & 0x7F); /* Mask PM bit if present for safety before conv? */
        t->day    = bcd2bin(t->day);
        t->month  = bcd2bin(t->month);
        t->year   = bcd2bin(t->year);
    }

    /* Convert 12 hour to 24 hour if necessary */
    if (!(registerB & 0x02) && (t->hour & 0x80)) {
        t->hour = ((t->hour & 0x7F) + 12) % 24;
    }

    /* Calculate full year */
    /* Assume 20th century if > 90? Or just assume 2000s for now */
    t->year += 2000;
}
